File: runtime/go/boundary.go
package shani

import "fmt"

// Capability grants the right to execute exactly one action (SPEC §7.2).
//
// Capabilities are single-use. Agents MUST NOT cache or reuse Capabilities.
type Capability struct {
	ado *AuthorizedDecisionObject
	used bool
}

// ADO returns the underlying ADO.
func (c *Capability) ADO() *AuthorizedDecisionObject {
	return c.ado
}

// Execute runs action gated by this capability (single-use).
func (c *Capability) Execute(action func(*AuthorizedDecisionObject) error) error {
	if c.used {
		return fmt.Errorf("capability already used — capabilities are single-use")
	}
	c.used = true
	return action(c.ado)
}

// ExecutionBoundary enforces that execution only happens through a valid ADO.
type ExecutionBoundary struct {
	evaluator *ShaniEvaluator
}

// NewExecutionBoundary creates a boundary backed by the given evaluator.
func NewExecutionBoundary(e *ShaniEvaluator) *ExecutionBoundary {
	return &ExecutionBoundary{evaluator: e}
}

// IssueCapability verifies an ADO and returns a Capability.
//
// Verifies signature and expiry, then consumes the nonce.
func (b *ExecutionBoundary) IssueCapability(ado *AuthorizedDecisionObject) (*Capability, error) {
	if err := b.evaluator.VerifyADO(ado); err != nil {
		return nil, fmt.Errorf("ADO verification: %w", err)
	}
	if err := b.evaluator.RegisterExecuted(ado); err != nil {
		return nil, fmt.Errorf("ADO registration: %w", err)
	}
	return &Capability{ado: ado}, nil
}

File: runtime/rust/src/boundary.rs
//! Execution Boundary — enforces that no action runs without a valid ADO (SPEC §7)

use crate::error::ShaniError;
use crate::evaluator::ShaniEvaluator;
use crate::schemas::decision::AuthorizedDecisionObject;

/// A Capability grants the right to execute exactly one action.
///
/// Capabilities are non-clonable and consumed on use.
/// Agents MUST NOT cache or reuse Capabilities (SPEC §7.2).
pub struct Capability {
    ado: AuthorizedDecisionObject,
}

impl Capability {
    /// Execute an action gated by this capability.
    ///
    /// The closure receives the ADO for context. Returns the action result.
    pub fn execute<F, T>(self, action: F) -> T
    where
        F: FnOnce(&AuthorizedDecisionObject) -> T,
    {
        action(&self.ado)
    }

    pub fn ado(&self) -> &AuthorizedDecisionObject {
        &self.ado
    }
}

/// Enforces that execution only happens through a valid, verified ADO.
pub struct ExecutionBoundary<'e> {
    evaluator: &'e mut ShaniEvaluator,
}

impl<'e> ExecutionBoundary<'e> {
    pub fn new(evaluator: &'e mut ShaniEvaluator) -> Self {
        Self { evaluator }
    }

    /// Verify an ADO and issue a Capability.
    ///
    /// Verifies signature and expiry, then consumes the nonce to prevent replay.
    pub fn issue_capability(
        &mut self,
        ado: AuthorizedDecisionObject,
    ) -> Result<Capability, ShaniError> {
        self.evaluator.verify_ado(&ado)?;
        self.evaluator.register_executed(&ado)?;
        Ok(Capability { ado })
    }
}
